src/common/hist_builder.h

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright 2017-2021 by Contributors
+ * \file hist_builder.h
+ */
+#ifndef XGBOOST_COMMON_HIST_BUILDER_H_
+#define XGBOOST_COMMON_HIST_BUILDER_H_
+
+#include <algorithm>
+#include <vector>
+#include "hist_util.h"
+#include "../data/gradient_index.h"
+
+#if defined(XGBOOST_MM_PREFETCH_PRESENT)
+  #include <xmmintrin.h>
+  #define PREFETCH_READ_T0(addr) _mm_prefetch(reinterpret_cast<const char*>(addr), _MM_HINT_T0)
+#elif defined(XGBOOST_BUILTIN_PREFETCH_PRESENT)
+  #define PREFETCH_READ_T0(addr) __builtin_prefetch(reinterpret_cast<const char*>(addr), 0, 3)
+#else  // no SW pre-fetching available; PREFETCH_READ_T0 is no-op
+  #define PREFETCH_READ_T0(addr) do {} while (0)
+#endif  // defined(XGBOOST_MM_PREFETCH_PRESENT)
+
+namespace xgboost {
+namespace common {
+
+struct Prefetch {
+ public:
+  static constexpr size_t kCacheLineSize = 64;
+  static constexpr size_t kPrefetchOffset = 10;
+
+ private:
+  static constexpr size_t kNoPrefetchSize =
+      kPrefetchOffset + kCacheLineSize /
+      sizeof(decltype(GHistIndexMatrix::row_ptr)::value_type);
+
+ public:
+  static size_t NoPrefetchSize(size_t rows) {
+    return std::min(rows, kNoPrefetchSize);
+  }
+
+  template <typename T>
+  static constexpr size_t GetPrefetchStep() {
+    return Prefetch::kCacheLineSize / sizeof(T);
+  }
+};
+
+template<typename FPType, bool do_prefetch,
+         typename BinIdxType, bool is_root,
+         bool any_missing>
+void BuildHistKernel(const std::vector<GradientPair>& gpair,
+                     const uint32_t* rows,
+                     const uint32_t row_begin,
+                     const uint32_t row_end,
+                     const GHistIndexMatrix& gmat,
+                     const BinIdxType* numa_data,
+                     uint16_t* nodes_ids,
+                     std::vector<std::vector<FPType>>* p_hists,
+                     const uint16_t* mapping_ids) {
+  const size_t size = row_end - row_begin;
+  const float* pgh = reinterpret_cast<const float*>(gpair.data());
+  const BinIdxType* gradient_index = numa_data;
+  const size_t* row_ptr =  gmat.row_ptr.data();
+  const uint32_t* offsets = gmat.index.Offset();
+  const size_t n_features = row_ptr[1] - row_ptr[0];
+  const uint32_t two {2};  // Each element from 'gpair' and 'hist' contains
+                           // 2 FP values: gradient and hessian.
+                           // So we need to multiply each row-index/bin-index by 2
+                           // to work with gradient pairs as a singe row FP array
+  std::vector<std::vector<FPType>>& hists = *p_hists;
+
+  for (size_t i = row_begin; i < row_end; ++i) {
+    const size_t ri = is_root ? i : rows[i];
+    const size_t icol_start = any_missing ? row_ptr[ri] : ri * n_features;
+    const size_t icol_end =  any_missing ? row_ptr[ri+1] : icol_start + n_features;
+    const size_t row_size = icol_end - icol_start;
+    const size_t idx_gh = two * ri;
+    const uint32_t nid = is_root ? 0 : mapping_ids[nodes_ids[ri]];
```

Comment:
It's unclear to me what is this mapping doing.

Reply:
This is simple mapping of the real node id (in the currently build tree) to node number on current depth (`0 <= nid < 2^current_depth`). This mapping allows to re-use allocated histograms.